src/main/java/com/cloudbees/jenkins/support/threaddump/HighLoadCpuChecker.java

```diff
@@ -0,0 +1,128 @@
+
+/**
+ * PeriodicWork to check when there is a high load in the instance.
+ *
+ * Only checking high CPU usage for the moment, but it can be used
+ * to generate thread dumps in high heap memory consumption.
+ */
+@Extension
+public class HighLoadCpuChecker extends PeriodicWork {
+
+    /**
+     * Recurrence period to check high cpu load consumption. Thread dumps after there are CONSECUTIVE_HIGH_CPU
+     * in the RECURRENCE_PERIOD_SEC
+     */
+    public static final int RECURRENCE_PERIOD_SEC =
+            Integer.getInteger(HighLoadCpuChecker.class.getName()+ ".RECURRENCE_PERIOD_SEC", 600);
+
+    /**
+     * Consecutive high CPUs to take a thread dump
+     */
+    public static final int HIGH_CPU_CONSECUTIVE_TIMES =
+            Integer.getInteger(HighLoadCpuChecker.class.getName()+ ".HIGH_CPU_CONSECUTIVE_TIMES", 3);;
+
+    /**
+     * This is the CPU usage threshold. Determinate de percentage of the total CPU used across all the
+     * cores available
+     */
+    public static final Double CPU_USAGE_THRESHOLD =
+            new Double(System.getProperty(HighLoadCpuChecker.class.getName() + ".CPU_USAGE_THRESHOLD", "0.80"));
+
+    /**
+     * Limit the number of thread dumps to retain on high cpu
+     */
+    public static final int HIGH_CPU_THEAD_DUMPS_TO_RETAIN =
+            Integer.getInteger(HighLoadCpuChecker.class.getName()+ ".HIGH_CPU_THEAD_DUMPS_TO_RETAIN", 5);
+
+    /**
+     * Thread dumps generated on high CPU load are stored in $JENKINS_HOME/high-load/cpu
+     **/
+    protected final FileListCap logs = new FileListCap(new File(Jenkins.getInstance().getRootDir(),"high-load/cpu"), HIGH_CPU_THEAD_DUMPS_TO_RETAIN);
+
+    private final SimpleDateFormat format = new SimpleDateFormat("yyyyMMdd-HHmmss.SSS");
+    {
+        format.setTimeZone(TimeZone.getTimeZone("UTC"));
+    }
+
+    private int countConsequtivePositives = 0;
+
+    @Override
+    public long getRecurrencePeriod() {
+        return TimeUnit.SECONDS.toMillis(RECURRENCE_PERIOD_SEC);
+    }
+
+    @Override
+    protected void doRun() throws Exception {
+        Double cpuLoad;
+        try {
+            VMMetricProviderImpl vmMetricProvider = ExtensionList.lookup(MetricProvider.class).get(VMMetricProviderImpl.class);
+            MetricSet vmProviderMetricSet = vmMetricProvider.getMetricSet();
+            Gauge cpu = (Gauge) vmProviderMetricSet.getMetrics().get("vm.cpu.load");
+            cpuLoad = new Double(cpu.getValue().toString());
+        } catch (NullPointerException nullPointerException) {
+            LOGGER.log(WARNING, "Support Core plugin can't generate thread dumps. Metrics plugin does not seem to be available", nullPointerException);
+            return;
+        }
+        if (Double.compare(Runtime.getRuntime().availableProcessors() * CPU_USAGE_THRESHOLD, cpuLoad) < 0) {
+            countConsequtivePositives++;
+            if (countConsequtivePositives >= HIGH_CPU_CONSECUTIVE_TIMES) {
+                    countConsequtivePositives = 0;
```

Comment:
Note this would happen at the 4th time, I think.

Reply:
@varyvol Why do you say that? I am using `>= HIGH_CPU_CONSECUTIVE_TIMES `. I just debugged it and it happens at the 3rd time. Am I missing something?

Reply:
Ah right, because you are increasing it before asking. Forget about my comment!